python/graphstorm/run/launch.py

```diff
@@ -0,0 +1,836 @@
+    )
+
+
+def wrap_cmd_with_local_envvars(cmd: str, env_vars: str) -> str:
+    """Wraps a CLI command with desired env vars with the following properties:
+    (1) env vars persist for the entire `cmd`, even if it consists of multiple "chained" commands like:
+        cmd = "ls && pwd && python run/something.py"
+    (2) env vars don't pollute the environment after `cmd` completes.
+
+    Example:
+        >>> cmd = "ls && pwd"
+        >>> env_vars = "VAR1=value1 VAR2=value2"
+        >>> wrap_cmd_with_local_envvars(cmd, env_vars)
+        "(export VAR1=value1 VAR2=value2; ls && pwd)"
+
+    Args:
+        cmd:
+        env_vars: A string containing env vars, eg "VAR1=val1 VAR2=val2"
+
+    Returns:
+        cmd_with_env_vars:
+
+    """
+    # use `export` to persist env vars for entire cmd block. required if udf_command is a chain of commands
+    # also: wrap in parens to not pollute env:
+    #     https://stackoverflow.com/a/45993803
+    return f"(export {env_vars}; {cmd})"
+
+
+def wrap_cmd_with_extra_envvars(cmd: str, env_vars: list) -> str:
+    """Wraps a CLI command with extra env vars
+
+    Example:
+        >>> cmd = "ls && pwd"
+        >>> env_vars = ["VAR1=value1", "VAR2=value2"]
+        >>> wrap_cmd_with_extra_envvars(cmd, env_vars)
+        "(export VAR1=value1 VAR2=value2; ls && pwd)"
+
+    Args:
+        cmd:
+        env_vars: A list of strings containing env vars, e.g., ["VAR1=value1", "VAR2=value2"]
+
+    Returns:
+        cmd_with_env_vars:
+    """
+    env_vars = " ".join(env_vars)
+    return wrap_cmd_with_local_envvars(cmd, env_vars)
+
+
+g_monitor_file = None
+g_group_id = 0
+
+def update_udf_command(udf_command, args):
+    """ Update udf_command with arguments from args.
+
+        The arguments to update includes:
+        1. ip-config
+        2. part-config
+    """
+    udf_command.append("--ip-config")
+    udf_command.append(args.ip_config)
+
+    udf_command.append("--part-config")
+    udf_command.append(args.part_config)
+
+    return udf_command
+
+def get_available_port(ip):
+    """Get available port with specified ip."""
+    import socket
+
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    for port in range(1234, 65535):
+        try:
+            sock.connect((ip, port))
+        except:
+            return port
+    raise RuntimeError("Failed to get available port for ip~{}".format(ip))
+
+
+def submit_jobs(args, udf_command):
+    """Submit distributed jobs (server and client processes) via ssh"""
+    servers_cmd = []
+    clients_cmd = []
+    hosts = []
+    thread_list = []
+    server_count_per_machine = 0
+
+    args.workspace = os.path.abspath(args.workspace)
+    # Get the IP addresses of the cluster.
+    # The path to the ip config file can be a absolute path or
+    # a relative path to the workspace
+    ip_config = args.ip_config if os.path.isabs(args.ip_config) else \
+        os.path.join(args.workspace, args.ip_config)
+    args.ip_config = ip_config
```

Comment:
is it right to change the argument value?

Reply:
ip_config and part_config will be latter passed to train or inference script.
Here I changed it into absolute path.